Assets/Scripts/Crab.cs

```diff
@@ -0,0 +1,154 @@
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        
+    }
+
+    private void FixedUpdate()
+    {
+        float vX = _moveSpeed;
+
+        if(_facingDirection == LEFT)
+        {
+            vX = -_moveSpeed;
+        }
+
+
+        _rb2d.velocity = new Vector2(vX, 0);
+
+        if (IsPlayerClose())
+        {
+            MoveTowardsPlayer();
+        }
+
+        if(IsPlayerInAttackRange())
+        {
+            Attack();
+        }
+
+        if (IsHittingWall() || IsAboutToFall())
+        {
+            ChangeFacingDirection();
+        }
+    }
+
+    private void ChangeFacingDirection()
+    {
+        Vector3 currentScale = transform.localScale;
+        currentScale.x = -transform.localScale.x;
+        transform.localScale = currentScale;
+
+        _facingDirection = _facingDirection == LEFT ? RIGHT : LEFT;
+    }
+
+    private void ChangeFacingDirection(string direction)
+    {
+        if(_facingDirection != direction)
+        {
+            ChangeFacingDirection();
+        }
+    }
+
+    private bool IsHittingWall()
+    {
+        float castDist = _facingDirection == LEFT ? -_baseHorizontalCastDist : _baseHorizontalCastDist;
+
+        Vector3 targetPos = _castPos.position;
+        targetPos.x += castDist;
+
+        Debug.DrawLine(_castPos.position, targetPos, Color.green);
+
+        bool val = Physics2D.Linecast(_castPos.position, targetPos, 1 << LayerMask.NameToLayer("Ground"));
+
+        return val;
+    }
+
+    private bool IsAboutToFall()
+    {
+        Vector3 targetPos = _castPos.position;
+        targetPos.y -= _baseVerticalCastDist;
+
+        Debug.DrawLine(_castPos.position, targetPos, Color.green);
+
+        bool val = !Physics2D.Linecast(_castPos.position, targetPos, 1 << LayerMask.NameToLayer("Ground"));
+
+        return val;
+    }
+
+    private bool IsPlayerClose()
+    {
+        // crab and player are on different levels vertically so shouldn't follow even if it's close
+        var verticalFollowDist = 0.6f;
+        if (Math.Abs(_playersTransform.position.y - _rb2d.transform.position.y) > verticalFollowDist)
+        {
+            return false;
+        }
+
+        return Vector2.Distance(_playersTransform.position, _rb2d.transform.position) < playerFollowDist;
+    }
+
+    private void MoveTowardsPlayer()
+    {
+        ChangeFacingDirection(_rb2d.transform.position.x < _playersTransform.position.x ? RIGHT : LEFT);
+    }
+
+    private bool IsPlayerInAttackRange()
+    {
+        var attackRange = 0.6f;
+        return Vector2.Distance(_playersTransform.position, _rb2d.transform.position) < attackRange;
+    }
+
+    private void Attack()
+    {
+        print(_anim.GetCurrentAnimatorClipInfo(0)[0].clip.name);
+        if (_anim.GetCurrentAnimatorClipInfo(0)[0].clip.name.Contains("Attack"))
+        {
+            return;
+        }
+        int attackNumber = UnityEngine.Random.Range(1, 4);
+        _anim.Play($"Enemy Attack {attackNumber}");
```

Comment:
This is okay for this type of enemy. But it would be nice to have some control over the attack rate (ex. add timer so we could control time between attacks)